internal/kubernetes/manager.go

```diff
@@ -3,111 +3,72 @@ package kubernetes_manager
 import (
 	"context"
 
+	"github.com/wavefrontHQ/wavefront-operator-for-kubernetes/internal/util"
+
 	"k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/api/meta"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
-	"k8s.io/apimachinery/pkg/types"
-	"k8s.io/apimachinery/pkg/util/json"
-	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-type KubernetesManager interface {
-	ApplyResources(resourceYamls []string, filterObject func(*unstructured.Unstructured) bool) error
-	DeleteResources(resourceYamls []string) error
-}
+type Client interface {
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object) error
 
-func NewKubernetesManager(mapper meta.RESTMapper, dynamicClient dynamic.Interface) (KubernetesManager, error) {
-	return &kubernetesManager{
-		RestMapper:    mapper,
-		DynamicClient: dynamicClient,
-	}, nil
+	Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error
+	Patch(ctx context.Context, obj client.Object, patch client.Patch, opts ...client.PatchOption) error
+	Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error
 }
 
-type kubernetesManager struct {
-	RestMapper    meta.RESTMapper
-	DynamicClient dynamic.Interface
+type KubernetesManager struct {
+	objClient Client
 }
 
-func (km kubernetesManager) ApplyResources(resourceYAMLs []string, filterObject func(*unstructured.Unstructured) bool) error {
-	var dynamicClient dynamic.ResourceInterface
+func NewKubernetesManager(objClient Client) *KubernetesManager {
+	return &KubernetesManager{objClient: objClient}
+}
 
-	for _, resource := range resourceYAMLs {
+func (km *KubernetesManager) ApplyResources(resourceYAMLs []string, exclude func(*unstructured.Unstructured) bool) error {
+	var resourceDecoder = yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	for _, resourceYAML := range resourceYAMLs {
 		object := &unstructured.Unstructured{}
-		var resourceDecoder = yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
-		_, gvk, err := resourceDecoder.Decode([]byte(resource), nil, object)
+		_, gvk, err := resourceDecoder.Decode([]byte(resourceYAML), nil, object)
 		if err != nil {
 			return err
 		}
 
-		if filterObject(object) {
+		if exclude(object) {
 			continue
 		}
 
-		mapping, err := km.RestMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		var getObj unstructured.Unstructured
+		getObj.SetGroupVersionKind(*gvk)
+		err = km.objClient.Get(context.Background(), util.ObjKey(object.GetNamespace(), object.GetName()), &getObj)
+		if errors.IsNotFound(err) {
+			err = km.objClient.Create(context.Background(), object)
+		} else if err == nil {
+			var diffObj unstructured.Unstructured
```

Comment:
If we were to diff against `getObj`, it could look like we're trying to reset immutable fields because they aren't set in the YAML that we're applying. Kubernetes of course gives us an error if we try to reset immutable fields. So what we need to do is diff against "nothing" so that only the fields we're trying to apply show up in the diff.